lib/bundler/definition.rb

```diff
@@ -472,28 +469,34 @@ def validate_runtime!
     end
 
     def validate_ruby!
-      return unless ruby_version
-
-      if diff = ruby_version.diff(Bundler::RubyVersion.system)
-        problem, expected, actual = diff
-
-        msg = case problem
-              when :engine
-                "Your Ruby engine is #{actual}, but your Gemfile specified #{expected}"
-              when :version
-                "Your Ruby version is #{actual}, but your Gemfile specified #{expected}"
-              when :engine_version
-                "Your #{Bundler::RubyVersion.system.engine} version is #{actual}, but your Gemfile specified #{ruby_version.engine} #{expected}"
-              when :patchlevel
-                if !expected.is_a?(String)
-                  "The Ruby patchlevel in your Gemfile must be a string"
-                else
-                  "Your Ruby patchlevel is #{actual}, but your Gemfile specified #{expected}"
-                end
-        end
-
-        raise RubyVersionMismatch, msg
-      end
+      return unless ruby_version = self.ruby_version || locked_ruby_version_object
+
+      return unless diff = ruby_version.diff(Bundler::RubyVersion.system)
+      problem, expected, actual = diff
+
+      implicit_ruby_version = self.ruby_version.nil?
+      specified = implicit_ruby_version ? "bundled is locked to" : "Gemfile specified"
+
+      msg = case problem
+            when :engine
+              "Your Ruby engine is #{actual}, but your #{specified} #{expected}"
+            when :version
+              return if implicit_ruby_version && ruby_version.to_gem_version_with_patchlevel < Bundler::RubyVersion.system.to_gem_version_with_patchlevel
+              "Your Ruby version is #{actual}, but your #{specified} #{expected}"
+            when :engine_version
+              "Your #{Bundler::RubyVersion.system.engine} version is #{actual}, but your #{specified} #{ruby_version.engine} #{expected}"
+            when :patchlevel
+              if !expected.is_a?(String)
+                "The Ruby patchlevel in your Gemfile must be a string"
+              else
+                "Your Ruby patchlevel is #{actual}, but your #{specified} #{expected}"
+              end
+      end
+      msg += ". This may cause issues if there are any gems that were depending on that ruby version." if implicit_ruby_version
+
+      raise RubyVersionMismatch, msg if !implicit_ruby_version || Bundler.feature_flag.bundler_2_mode?
```

Comment:
Oh man, I completely misunderstood this question, sorry. How about `ruby.strict_lock`?